feat: allow specify url in release assets

awcjack opened the pull request:
This solve https://github.com/semantic-release/gitlab/issues/154

hvanoch commented:
Is there something holding back this MR? 
This would be helpful for us too, so if I can assist somewhere getting this merged.

JonasSchubert commented on lib/publish.js:
```diff
@@ -36,47 +36,53 @@ module.exports = async (pluginConfig, context) => {
-          return;
+        const _url = asset.url
+        if (_url) {
+          assetsList.push({ label, url: _url, type, filepath });
```
Unfortunately this will not work as intended.

I wanted to pull these changes into the PR #343 and added an unit test. This is the test I had to write to make it work:
```javascript

test.serial('Publish a release with an asset link', async t => {
  const cwd = 'test/fixtures/files';
  const owner = 'test_user';
  const repo = 'test_repo';
  const env = {GITLAB_TOKEN: 'gitlab_token'};
  const nextRelease = {gitHead: '123', gitTag: 'v1.0.0', notes: 'Test release note body'};
  const options = {repositoryUrl: `https://gitlab.com/${owner}/${repo}.git`};
  const encodedRepoId = encodeURIComponent(`${owner}/${repo}`);
  const encodedGitTag = encodeURIComponent(nextRelease.gitTag);
  const link = {
    path: 'file.css',
    label: 'README.md',
    type: 'other',
    url: 'https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md',
  };
  const assets = [link];
  const gitlab = authenticate(env)
    .post(`/projects/${encodedRepoId}/releases`, {
      tag_name: nextRelease.gitTag,
      description: nextRelease.notes,
      assets: {
        links: [
          {
            name: 'README.md',
            url: `https://gitlab.com/${owner}/${repo}/https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md`,
            link_type: 'other',
          },
        ],
      },
    })
    .reply(200);

  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});

  t.is(result.url, `https://gitlab.com/${owner}/${repo}/-/releases/${encodedGitTag}`);
  t.deepEqual(t.context.log.args[0], ['Published GitLab release: %s', nextRelease.gitTag]);
  t.true(gitlab.isDone());
});
```

- I had to provide a non empty **path** to a file, otherwise [getAssets](https://github.com/awcjack/gitlab/blob/06915a124346ae8ed0239ffd8f9d18369e579033/lib/publish.js#L33) fails

```bash
  publish › Publish a release with an asset link

  test\publish.test.js:161

   160:
   161:   const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});
   162:

  Rejected promise returned by test. Reason:

  TypeError {
    message: 'Cannot read properties of undefined (reading \'0\')',
  }

  › getPath (node_modules/dir-glob/index.js:8:22)
  › node_modules/dir-glob/index.js:55:50
  › Array.map (<anonymous>)
  › module.exports (node_modules/dir-glob/index.js:54:51)
  › null.<anonymous> (lib/glob-assets.js:17:36)
  › Array.map (<anonymous>)
  › null.module.exports (lib/glob-assets.js:13:18)
  › null.module.exports (lib/publish.js:34:33)
  › test/publish.test.js:161:24

  ─
```

Additionally the created URL will be joined with the URI to the repo the release was created in and the provided URL will be appended, resulting in something like this: `https://gitlab.com/test_user/test_repo/https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md`

JonasSchubert replied:
Maybe we can filter assets with `url` before we call `getAssets` and handle that beforehand.

JonasSchubert commented on lib/publish.js:
```diff
@@ -36,47 +36,53 @@ module.exports = async (pluginConfig, context) => {
-        } catch (_) {
-          logger.error('The asset %s cannot be read, and will be ignored.', path);
-          return;
+        const _url = asset.url
```
I think a small remark in the `README.md` would be helpful, to point users to this feature. E.g.:

```markdown
// ...

{
  "plugins": [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    ["@semantic-release/gitlab", {
      "gitlabUrl": "https://custom.gitlab.com",
      "assets": [
        {"path": "dist/asset.min.css", "label": "CSS distribution"},
        {"path": "dist/asset.min.js", "label": "JS distribution", "type": "generic_package"},
        {"url": "https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md"}
      ]
    }],
  ]
}

// ...

#### assets

// ...

| `url`   | This provides the ability to add links to releases. E.g. URLs to container images. | -                                    |
```

awcjack commented:
forget created pull request
I thought I just test in my repo and haven't touch the repo few months lol

JonasSchubert commented on lib/publish.js:
```diff
@@ -30,53 +30,78 @@ module.exports = async (pluginConfig, context) => {
-          throw error;
+        let _url = asset.url;
+        if (_url) {
+          // Allow loading variable from process.env load CI_JOB_ID for attaching CI artifact
```
Nice you considered this, too 👍

fgreinacher replied:
I understand the idea behind this, but I don't like how this new propery is handled in a very special way. We already have a templating mechanism (via lodash) and I don't want to introduce another one. I'd suggest we switch to that for the URL and create a follow-up issue for using env vars in templates.

JonasSchubert commented on README.md:
```diff
@@ -72,7 +73,8 @@ Can be a [glob](https://github.com/isaacs/node-glob#glob-primer) or and `Array`
 | Property | Description                                                                                                 | Default                              |
 | -------- | ----------------------------------------------------------------------------------------------------------- | ------------------------------------ |
-| `path`   | **Required.** A [glob](https://github.com/isaacs/node-glob#glob-primer) to identify the files to upload.    | -                                    |
+| `url`    | **path or url is required.** This provides the ability to add links to releases. E.g. URLs to container images.                           | -                                    |
```
I would put the path up as this may still be the most commonly used feature; to weight it higher. @awcjack What do you think?

JonasSchubert commented on lib/publish.js:
```diff
@@ -30,53 +30,78 @@ module.exports = async (pluginConfig, context) => {
+            for (const matchedVariable of matchedVariables) {
+              const variableName = matchedVariable.slice(2, -1);
+              if (process.env[variableName]) {
+                _url = _url.replace(matchedVariable, process.env[variableName]);
```
This will only find the first usage.
I do not know if somebody might do this, but if one variable might be used multiple times, it will not be replaced.

Maybe:
```suggestion
                _url = _url.replace(new RegExp('\\' + matchedVariable, 'g), process.env[variableName]);
```
The `\\` are needed to escape `$`.

@awcjack What do you think?

awcjack replied:
Fixed the global flag first

awcjack replied:
Using escapeRegExp function now for escaping all special symbol in regexp

awcjack replied:
```
function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}
```

JonasSchubert reviewed (changes requested):
@fgreinacher  What do you think?

awcjack commented on test/publish.test.js:
```diff
@@ -247,3 +247,80 @@ test.serial('Publish a release with missing release notes', async t => {
+  const link = {
+    label: 'README.md',
+    type: 'other',
+    url: 'https://gitlab.com/gitlab-org/${repo}/-/blob/master/README.md', // eslint-disable-line no-template-curly-in-string
```
After adding the backslash before dollar sign, the linter throw error like below
```
  test/publish.test.js:302:10
  ✖  302:10  Unexpected template string expression.  no-template-curly-in-string
  ✖  302:41  Delete \                                prettier/prettier
  ✖  302:41  Unnecessary escape character: \$.       no-useless-escape
```

JonasSchubert replied:
Sorry! My mistake, we have to use different quotes.

awcjack replied:
Done

awcjack commented on lib/publish.js:
```diff
@@ -30,53 +30,78 @@ module.exports = async (pluginConfig, context) => {
+            }
+          }
+
+          assetsList.push({label, rawUrl: _url, type, filepath});
```
Not so sure how it works but keeping `filepath` here for providing flexibility to use [Permanent links to release assets feature](https://docs.gitlab.com/ee/user/project/releases/index.html#permanent-links-to-release-assets)

JonasSchubert reviewed (approved):
Only merge issues to fix and then it looks good for me. (sorry for the late reply...)
@fgreinacher what do you think?

fgreinacher reviewed (changes requested):
Code looks very good, I have some concerns with the templating though.

And we need a rebase 😀

awcjack commented:
Rebased (ran test for checking no merge error) and removed the variable code  
I will create another merge request after updating the variable code to lodash template later
@fgreinacher

fgreinacher commented on README.md:
```diff
@@ -95,7 +96,8 @@ Can be a [glob](https://github.com/isaacs/node-glob#glob-primer) or and `Array`
 | Property | Description                                                                                                 | Default                              |
 | -------- | ----------------------------------------------------------------------------------------------------------- | ------------------------------------ |
-| `path`   | **Required.** A [glob](https://github.com/isaacs/node-glob#glob-primer) to identify the files to upload.    | -                                    |
+| `path`   | **path or url is required.** A [glob](https://github.com/isaacs/node-glob#glob-primer) to identify the files to upload.    | -                                    |
+| `url`    | **path or url is required.** This provides the ability to add links to releases. E.g. URLs to container images.                           | -                                    |
```
```suggestion
| `path`   | **Required**, unless `url` is set. A [glob](https://github.com/isaacs/node-glob#glob-primer) to identify the files to upload.    | -                                    |
| `url`.     | Alternative to setting `path` this provides the ability to add links to releases, e.g. URLs to container images.                           | -                                    |
```

fgreinacher commented:
@awcjack Thanks a lot for your contribution, greatly appreciated!
